### DISCUSSION TITLE
Pourquoi les composants sont des interfaces?

### DISCUSSION THREAD

**Discussion by nasso** [TempoDev/Arcade-Toulouse#36] [Questions]
Pour moi, un composant n'est rien d'autre qu'une structure de donnée. Il n'y a pas d'"action" qu'on puisse faire ou faire faire un composant.

Pourquoi les composants `ISprite`, `IAsciiSprite`, `IText` etc... sont-ils des interfaces plutôt que de simples structures?

**Comment by Pebloop** [comment]
Parce que:
* Jérôme veut des interfaces
* ça permet une implémentation plus générale : Le sprite en lui même c'est un tableau de pixel mais la façon dont tu implémentes ce tableau est 100% libre, tu devras juste respecter une suite de méthodes `getPixelAt(int index)`, `getWidth() const`, `getHeigth() const`, ...

Et comme tu ne peux pas créer une interface (vu que c'est abstrait) tu dois aussi créer une factory pour chacune d'entres elles.

**Reply by nasso** [reply]
Je trouve ça bien trop complexe pour le coup, ya pas 36 façon d'implémenter un tableau de pixel. Ça pose aussi de très gros soucis de performance pour pas grand chose, par exemple si ont peut lire les pixels que un par un.

On a une implémentation très simple [sur mon fork](https://github.com/nasso/Arcade-Toulouse/blob/master/src/IScene.hpp#L34) qui je trouve offre justement beaucoup de liberté de par sa simplicité (possibilité de `memcpy` les pixels, de les passer directement à la lib graphique etc...).

**Reply by Pebloop** [reply]
Y a pleins de façons de l'implémenter tu veux dire, déjà suppose que en fait je veux plutôt faire un double tableau ? 
Et parle pas de 'mais c'est pas opti' parce que c'est pas le sujet. Y a des gens qui veulent juste un programme qui marche et il seront déja content.
Là tu imposes ta structure à tout le monde. Et c'est justement ce qu'il ne faut pas faire.

En ce qui concerne la lecture ça change rien parce que le core connait la classe derrière et peut donc faire ce qui lui chante.
Concernant la création, vu que tu es censé les générer au init (charger des sprites dans la loop n'est pas la plus opti des idées), à part un peut de chargement au début ça ne changera rien.

Maintenant avec
```cpp
setPixel(int x, int y);
setWidth(int width);
setHeight(int height);
```
tu peux tout faire du côté de IGame.
un static_cast du côté du core suffira.

**Reply by Pebloop** [reply]
PS, sur ton fork `Sprite` c'est une structure, or Jérôme a été assez clair je pense sur le "pas de void *". Donc ta struct doit au moins hériter d'un `IComponent`

**Reply by nasso** [reply]
J'ai pas de `void*` :p

**Reply by nasso** [reply]
> Y a pleins de façons de l'implémenter tu veux dire

Exactement, et comme ici on traite de donnée pure, choisir sa représentation fait partie de l'API commune. Une structure de donnée fait complètement partie d'une API au même titre qu'une interface: il n'y a pas d'implémentation de logique (pour faire le parallèle, c'est dans les `.h`/`.hpp` qu'on définie les structures et la représentation des données).

> Là tu imposes ta structure à tout le monde. Et c'est justement ce qu'il ne faut pas faire.

Je l'impose a personne, je fais un POC de mon côté sur quelque chose qu'on pourrait avoir. Ça permet d'évaluer la solution avant de la proposer. Jusque là j'ai trouvé cette solution comme étant la plus simple d'un point de vue utilisation et implémentation.

> En ce qui concerne la lecture ça change rien parce que le core connait la classe derrière et peut donc faire ce qui lui chante.

C'est surtout la lib graphique qui va devoir lire les pixels pour les afficher!

**Reply by Pebloop** [reply]
Ok, j'ai regardé ton truc plus en détail et je vois le truc, ton sprite c'est pas un composant.
Or autant le transform je peux comprendre, c'est un peu indispensable, et c'est la base, autant le sprite reste un composant, tu peux avoir une entitée sans sprite et tu ne peux pas commencer à tout mettre dans un IEntity parce que tu dois pouvoir créer des comps facultatifs pour ceux qui veulent aller plus loin, d'où une liste de comp lisible.

Concernant la lecture... j'ai peut-être zappé le fait que c'est pas le core qui interprète les composants ?
Dans ce cas je peux comprendre l'idée d'utiliser une struct commune.
Je parlais d'inposer parce que techniquement si tu proposes une struct tout le monde est obligé d'avoir **exactement** la même, donc ça laisse 0 liberté, mais c'est vrai que lire une image pixel par pixel c'est chaud.

**Reply by nasso** [reply]
Mon sprite est 100% un composant.

Imposer une interface est exactement pareil qu'imposer une structure.

**Reply by Pebloop** [reply]
Non, c'est pas du tout pareil, son implémentation reste libre, c'est le principe même d'une interface

**Reply by nasso** [reply]
Quand ton interface prend un int en paramètre ou retourne un int, tu impose un int a tout le monde.

Remplace "int" par "structure".
